### PR TITLE
Allow switching tool versions in workflow editor

### LINKED ISSUE #557: Selection of tool version in workflow editor
After installing multiple versions of a tool, it is not possible to select old versions when adding it as a step in the workflow editor

From @nsoranzo on [trello](https://trello.com/c/d7B3XHCR/1270-after-installing-multiple-versions-of-a-tool-it-is-not-possible-to-select-old-versions-when-adding-it-as-a-step-in-the-workflow-)

### PR DESCRIPTION
**Pull request by mvdbeek**
This will finally close https://github.com/galaxyproject/galaxy/issues/557 !

### CODE REVIEW

**Comment by mvdbeek** [comment]
This works nicely: If the input or output connections change, the connection is removed and the user can re-connect.

**Comment by gregvonkuster** [comment]
Awesome!  Thanks @mvdbeek ;)

**Comment by jmchilton** [comment]
@mvdbeek You are a wizard who packs so much awesome in such tiny commits.

**Comment by bgruening** [comment]
So cool! :beers: for Marius!

**Comment by nsoranzo** [comment]
@mvdbeek There's a conflict in `static/` and qunit tests are failing, can you take a look?

File: client/galaxy/scripts/mvc/workflow/workflow-forms.js [hunk 1]
```diff
@@ -37,7 +37,7 @@ define( [ 'utils/utils', 'mvc/form/form-view', 'mvc/tool/tool-form-base' ], func
                     text_disable    : 'Set at Runtime',
                     narrow          : true,
                     initial_errors  : true,
-                    sustain_version : true,
+                    sustain_version : false,
```
Comment (guerler) [inline client/galaxy/scripts/mvc/workflow/workflow-forms.js]: This line can be removed. The default is `false`.

File: client/galaxy/scripts/mvc/workflow/workflow-node.js [hunk 2]
```diff
@@ -185,6 +186,12 @@ define(['mvc/workflow/workflow-view-node'], function( NodeView ) {
             var nodeView = node.nodeView;
             this.tool_state = data.tool_state;
             this.config_form = data.config_form;
+            if (this.config_form) {
+                this.tool_version = this.config_form.version;
```
Comment (guerler) [inline client/galaxy/scripts/mvc/workflow/workflow-node.js]: Can be a one-liner: `this.tool_version = this.config_form && this.config_form.version;`.

File: client/galaxy/scripts/mvc/tool/tool-form-base.js [hunk 3]
```diff
@@ -156,7 +156,14 @@ define( [ 'utils/utils', 'utils/deferred', 'mvc/ui/ui-misc', 'mvc/form/form-view
                                 // queue model request
                                 self.deferred.reset();
                                 self.deferred.execute( function( process ) {
-                                    self._buildModel( process, { id : id, version : version } )
+                                    if ( options.hasOwnProperty( "workflow" ) ) {
```
Comment (guerler) [inline client/galaxy/scripts/mvc/tool/tool-form-base.js:159]: This should be moved out of the base class. I can help with this tho. Is this PR already functional?

Comment (mvdbeek) [reply client/galaxy/scripts/mvc/tool/tool-form-base.js:159]: Not entirely, I'm still struggling with getting the connections right when the number and type of output changes.

Comment (mvdbeek) [reply client/galaxy/scripts/mvc/tool/tool-form-base.js:159]: @guerler I think this is fully functional now, but I'm sure there are some optimizations to do. Feel free to push to this branch if you want.

File: client/galaxy/scripts/mvc/workflow/workflow-node.js [hunk 4]
```diff
@@ -183,8 +185,46 @@ define(['mvc/workflow/workflow-view-node'], function( NodeView ) {
...
+                delete node.output_terminals[unused_output];  // removes the output terminal
+            });
+            $.each( data.data_outputs, function( i, output ) {
+                if (!nodeView.outputViews[output.name]) {
+                    nodeView.addDataOutput(output);  // add data output if it does not yet exist
+                }
+            });
             this.tool_state = data.tool_state;
             this.config_form = data.config_form;
+            this.force_refesh = this.config_form && this.tool_version !== this.config_form.version;
```
Comment (nsoranzo) [inline client/galaxy/scripts/mvc/workflow/workflow-node.js]: s/force_refesh/force_refresh/ ? Also below.

File: client/galaxy/scripts/mvc/workflow/workflow-node.js [hunk 5]
```diff
@@ -154,10 +154,12 @@ define(['mvc/workflow/workflow-view-node'], function( NodeView ) {
             }
             this.name = data.name;
             this.config_form = data.config_form;
+            this.tool_version = this.config_form && this.config_form.version;
             this.tool_state = data.tool_state;
             this.errors = data.errors;
             this.tooltip = data.tooltip ? data.tooltip : "";
             this.annotation = data.annotation;
+            // this.output_terminals = {};  // removes outputs when switching tool versions
```
Comment (nsoranzo) [inline client/galaxy/scripts/mvc/workflow/workflow-node.js]: Leftover?

File: client/galaxy/scripts/mvc/tool/tool-form-base.js (hunk 3 above)
Comment (jmchilton) [reply client/galaxy/scripts/mvc/tool/tool-form-base.js:159]: @guerler Do you object to me merging as is so you can follow up with that change - or do you want this change made first?

**Comment by martenson** [comment]
@mvdbeek this needs dev merged in and client rebuild in this order, do you want me to do it?

**Comment by mvdbeek** [comment]
@martenson Sure, go ahead.

**Comment by martenson** [comment]
So in my testing I have FastQC Galaxy version 0.67 and 0.68 and switching in between them looks fine, however the default selected when adding to the WF canvas is 0.67, shouldn't it be the latest instead?

**Comment by martenson** [comment]
This implementation seems to select whatever revision was installed the last as the default.

**Comment by mvdbeek** [comment]
Hmm, is that caused by this PR? The default version should not be affected
here, but there were was a reports of similar behavior in
https://github.com/galaxyproject/galaxy/issues/557#issuecomment-213498153

**Comment by martenson** [comment]
@mvdbeek when selecting fastqc in standard analysis interface, I get the latest as default

**Comment by martenson** [comment]
Thanks a lot @mvdbeek !

File: client/galaxy/scripts/mvc/tool/tool-form-base.js (hunk 3 above)
Comment (guerler) [reply client/galaxy/scripts/mvc/tool/tool-form-base.js:159]: Thanks @mvdbeek. Yeah I can make those changes later.

**Pull request merged**